File: lua/protein-translation/protein-translation.lua
local proteins = {
    AUG = "Methionine",
    UUU = "Phenylalanine", UUC = "Phenylalanine",
    UUA = "Leucine", UUG = "Leucine",
    UCU = "Serine", UCC = "Serine", UCA = "Serine", UCG = "Serine",
    UAU = "Tyrosine", UAC = "Tyrosine",
    UGU = "Cysteine", UGC = "Cysteine",
    UGG = "Tryptophan",
    UAA = "STOP", UAG = "STOP", UGA = "STOP",
}

local codon = function(codon)
    assert(proteins[codon], "Unknown codon: "..codon)
    return proteins[codon]
end

local rna_strand = function(rna)
    local proteins = {}
    for i = 1, #rna, 3 do
        local cod = rna:sub(i, i+2)
        local protein = codon(cod)    -- allow error to propagate
        if protein == "STOP" then break end
        table.insert(proteins, protein)
    end
    return proteins
end

return { codon = codon, rna_strand = rna_strand }

File: nim/protein-translation/protein_translation.nim
proc codon2protein(codon: string): string =
  case codon
  of "AUG":                      "Methionine"
  of "UUU", "UUC":               "Phenylalanine"
  of "UUA", "UUG":               "Leucine"
  of "UCU", "UCC", "UCA", "UCG": "Serine"
  of "UAU", "UAC":               "Tyrosine"
  of "UGU", "UGC":               "Cysteine"
  of "UGG":                      "Tryptophan"
  of "UAA", "UAG", "UGA":        "STOP"
  else:
    raise newException(ValueError, "unknown codon")

proc translate*(rna: string): seq[string] =
  for i in countup(0, rna.len - 1, 3):
    var protein = codon2protein rna[i ..< i+3]
    if protein == "STOP": break
    result.add protein
